working-groups/unsafe-code-guidelines/notes/2019-03-14.md

```diff
@@ -0,0 +1,131 @@
+  - We described a [conservative version of "discriminant
+    elision"](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/enums.md#discriminant-elision-on-option-like-enums)
+    -- aka, "the option optimization" -- that people can rely on
+  - Layout of other enums is not defined
+- Unions tagged with `#[repr(C)]` behave like C unions. repr(rust)
+  unions are not specified -- in particular, some of the fields may
+  not be laid out at offset 0.
+- [Function
+  pointers](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/function-pointers.md)
+  are defined as being laid out the same way as their C counterparts,
+  though they cannot be NULL.
+- [Integers and
+  booleans](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/integers-floatingpoint.md)
+  and [SIMD vector
+  types](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/vectors.md)
+  are basically defined as C compatible; we did not revisit the "great bool debate".
+- [References and raw
+  pointers](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/pointers.md)
+  are largely specified as behaving as they do in rustc
+  today. However, we do not specify the layout of `&dyn (Trait1 +
+  Trait2)`, to leave room for multi-trait objects.
+
+We are currently discussing **validity invariants**. A **validity
+invariant** is an invariant that must "always" be maintained or else
+undefined behavior will result ([defining always more precisely is one
+of the things we are talking about][extent-of-always]). This permits
+the compiler to take advantage of these invariants for layout
+optimizations and other purposes.
+
+[extent-of-always]: https://github.com/rust-lang/unsafe-code-guidelines/blob/19f422384c5f8316235ac48917a59a6becbda99f/active_discussion/validity.md#extent-of-always
+
+- For the "boring stuff" (enums, structs, arrays), there is relatively
+  little controversy.  These invariants are largely created by
+  combinined other invariants, and gnzlbg plans to do a write-up on
+  these topics soon.
+- For integers, references, and unions, there are a wide variety of
+  opinions on the table:
+  - On the one hand, stronger invariants enable more layout optimizations,
+    as well as affecting things like [exhaustive matching][np].
+  - On the other, we want to enable common unsafe code patterns.
+  - Finally, we want to ensure that we have a feasible implementation
+    in miri that can do dynamic checking and detect UB.
+
+[np]: http://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/
+
+## Places we could use help
+
+- Writing up a summary that tries to clearly lay out the
+  options in the integer/reference threads thus far (many of those
+  issues are intertwined).
```

Comment:
Actually I thought about this some more, and came to the conclusion that maybe they are not that intertwined. For integers (and, by extension, raw pointers) the pattern we are worried about is
```
let buf: [u8; 1024] = mem::uninitialized();
something.read(&mut buf);
```
which is equivalent to what libc does when it lets external code initialize a struct (that consists solely of integers and raw pointers).

While references are involved (note the `&mut`), really the main thing here is having a *local variable* that is not initialized properly.

The discussions intertwine only once you consider
```
let mut buf = MaybeUninit::<[u8; 1024]>::uninitialized();
something.read(buf.get_mut());
// or, equivalently (with the subset of MaybeUninit that is going to be stabilized first)
something.read(&mut *buf.as_mut_ptr());
```
which we could allow either by declaring uninitialized integers valid, or by declaring references non-recursive for validity. We could also disallow this pattern.

Reply:
Without getting into the details here, I'm trying to decide what to do with this comment. I feel like this is quite a bit more detailed than the text at hand, but quite interesting. I don't quite think I want to fit into the summary but neither do I want to lose it. Perhaps for now I'll just add a link and we can decide how to cover it (if at all) in the meeting. =)

Reply:
(OK, I added a link.)